sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs

```diff
@@ -66,27 +67,89 @@ public static BinaryData Write(object model, ModelReaderWriterOptions? options =
+                }
 
-            if (options.Format == "J" || (options.Format == "W" && iModel.GetFormatFromOptions(options) == "J"))
+                return WriteEnumerable(options, enumerable);
+            }
+            else
             {
-                if (model is not IJsonModel<object> jsonModel)
+                if (ShouldUseJsonInterface(iModel, options))
+                {
+                    using (ModelWriter<object> writer = new ModelWriter<object>(GetJsonInterface(iModel, options), options))
+                    {
+                        return writer.ToBinaryData();
+                    }
+                }
+                else
                 {
-                    throw new FormatException($"The model {model.GetType().Name} does not support '{options.Format}' format.");
+                    return iModel.Write(options);
                 }
+            }
+        }
 
-                using (ModelWriter<object> writer = new ModelWriter<object>(jsonModel, options))
+        private static BinaryData WriteEnumerable(ModelReaderWriterOptions options, IEnumerable enumerable)
+        {
+            bool checkedElementFormat = false;
+            using MemoryStream stream = new MemoryStream();
+            using Utf8JsonWriter writer = new Utf8JsonWriter(stream);
+            writer.WriteStartArray();
+            foreach (var item in enumerable)
+            {
+                if (!checkedElementFormat)
                 {
-                    return writer.ToBinaryData();
+                    if (item is not IPersistableModel<object> itemAsModel)
+                    {
+                        throw new InvalidOperationException($"{item.GetType().Name} does not implement {nameof(IPersistableModel<object>)} or {nameof(IEnumerable<IPersistableModel<object>>)}");
+                    }
+                    if (!ShouldUseJsonInterface(itemAsModel, options))
+                    {
+                        throw new InvalidOperationException($"Writing a collection is only supported for 'J' format");
+                    }
+                    checkedElementFormat = true;
                 }
+                BinaryData itemData = Write(item, options);
+#if NET6_0_OR_GREATER
+                writer.WriteRawValue(itemData);
+#else
+                    using var doc = JsonDocument.Parse(itemData);
+                    JsonSerializer.Serialize(writer, doc.RootElement);
+#endif
+            }
+            writer.WriteEndArray();
+            writer.Flush();
+            return new BinaryData(stream.GetBuffer().AsMemory(0, (int)stream.Position));
+        }
+
+        private static void ReadList(BinaryData data, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type elementType, IList iList, ModelReaderWriterOptions options)
+        {
+            var converter = GetInstance(elementType);
```

Comment:
should tis be called readerProxy, not converter? Wen I was reading the code it confused me as I thought you are using JsonConverter here.